source/auth/auth.rst

```diff
@@ -250,6 +250,48 @@ or "SCRAM-SHA-256").
 The cache entry value MUST be either the ``saltedPassword`` parameter or the
 combination of the ``clientKey`` and ``serverKey`` parameters.
 
+Speculative Authentication via Handshake
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:since: 4.4
+
+The ``isMaster`` handshake supports a new argument, ``speculativeSaslStart``,
+provided as a BSON object. Clients specifying this argument to ``isMaster`` will
+speculatively include the first ``saslStart`` command of an authentication handshake.
+This command may be provided to the server in parallel with any standard request for
+supported authentication mechanisms. This would permit clients to merge their
+``saslStart`` message with their ``isMaster`` request, and receive the ``saslStart``
+reply with the ``isMaster`` reply.
+
+If the speculatively issued command succeeds, the client should proceed with the next
+step of the exchange. If the command fails, the client should determine whether the
+``isMaster`` response contains enough information to select another mechanism.
+
+Older servers will ignore this argument. New servers will participate in the standard
+authentication conversation if this argument is missing.
+
+An example conversation using ``speculativeSaslStart`` is as follows:
+
+| C: :javascript:`{ isMaster: 1, speculativeSaslStart: { mechanism: "SCRAM-SHA-256", db: "admin", payload: BinData(0, "biwsbj11c2VyLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdM"), options: { skipEmptyExchange: true }}, ... }`
+| S: :javascript:`{ isMaster: true, speculativeSaslStart: { conversationId: 1, payload: BinData(0,"cj1meWtvK2QybGJiRmdPTlJ2OXFreGRhd0xIbytWZ2s3cXZVT0tVd3VXTElXZzRsLzlTcmFHTUhFRSxzPXJROVpZM01udEJldVAzRTFURFZDNHc9PSxpPTEwMDAw"), done: false }, ... }`
+| C: :javascript:`{saslContinue: 1, conversationId: 1, payload: BinData(0, "Yz1iaXdzLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdMSG8rVmdrN3F2VU9LVXd1V0xJV2c0bC85U3JhR01IRUUscD1NQzJUOEJ2Ym1XUmNrRHc4b1dsNUlWZ2h3Q1k9")}`
+| S: :javascript:`{conversationId: 1, payload: BinData(0,"dj1VTVdlSTI1SkQxeU5ZWlJNcFo0Vkh2aFo5ZTA9"), done: true, ok: 1}`
+  
+
+Alternatively, the ``isMaster`` handshake supports another new argument,
+``speculativeAuthenticate``, provided as a BSON object. This argument provides an
+instance of the ``authenticate`` command with the ``MONGODB-X509`` mechanism.
+Servers accepting the request will respond with an ``isMaster`` reply containing the results
+in a field named ``speculativeAuthenticate``. Concurrent use of ``speculativeSaslStart``
+and ``speculativeAuthenticate`` will result in ``isMaster`` returning an error.  In this
+mode, authentication can occur in zero round trips and will be completed once ``isMaster``
+returns.
```

Comment:
The new paragraphs here don't actually instruct drivers to use these new `isMaster` options. Should we explicitly state that drivers MUST employ `speculativeAuthenticate` if the application has indicated it wishes to authenticate with `MONGODB-X509`? Likewise for requiring drivers to attempt `speculativeSaslStart` if the application indicates a SCRAM mechanism or no mechanism at all (in which case SCRAM-SHA-1 is used), per https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#defaults.

You may optionally want to note something to the effect of:

 *  drivers MUST NOT use `speculativeSaslStart` if an auth mechanism other than SCRAM-SHA-256 or SCRAM-SHA-1 has been indicated
 * drivers MUST NOT use `speculativeAuthenticate` if `MONGODB-X509` has not been indicated"

Reply:
I have added wording to this effect prior to the examples.